### DISCUSSION TITLE
EKMC model implementation

### DISCUSSION THREAD

**Discussion by RonanRiverie** [thelfer/tfel#462] [Q&A] [answered]
Hello again,

So, mfront is working perfectly on my computer on Windows together with CAST3M23 thanks to the help from last time. After, I tested few codes and now I am getting into the EKMC model which is the one to implement. Actually, I found on this forum that it has already been done here: https://github.com/thelfer/tfel/discussions/316. First of all, I just tried this model on a simple case with mfront and it seems to work fine. However, I have few questions about it and the implementations of the equations.

The equations of the strain aging model are as follows: 
![image](https://github.com/thelfer/tfel/assets/65529689/f5d363f7-0258-475a-a56e-6b5c8e28717a)

Here is your original code (with modifications just the name of the variables and I plotted the results in Python and it gives the same results):
```
@DSL Implicit;
@Behaviour EKMC;

@Algorithm NewtonRaphson_NumericalJacobian;
@PerturbationValueForNumericalJacobianComputation 1e-8;
@Epsilon 1e-14;
@Theta 1;

@Brick StandardElasticity;

@MaterialProperty strainrate de0;
@MaterialProperty real Ea;
@MaterialProperty real Va;
@MaterialProperty stress P1;
@MaterialProperty real P2;
@MaterialProperty real omega;
@MaterialProperty stress R0;
@MaterialProperty stress Q;
@MaterialProperty real b;
@MaterialProperty stress H;
@MaterialProperty real nta;
@MaterialProperty real alpha;

@StateVariable strain p;
p.setGlossaryName("EquivalentViscoplasticStrain");

@AuxiliaryStateVariable time ta;
ta.setEntryName("MaterialTime");

@LocalVariable time dta;

@LocalVariable strainrate de;

@LocalVariable stress sinh0;

@InitLocalVariables {
  constexpr auto kB = PhysicalConstants::kb;
  const auto T_theta = T + theta * dT;
  de = de0 * exp(-Ea / (kB * (T_theta)));
  sinh0 = kB * T_theta / Va;
}

@Integrator {
  const auto seq = sigmaeq(sig);
  Stensor n = Stensor(0);
  if(seq > 1.e-12){
    n = 1.5 * deviator(sig)/seq;
  }

  feel = deel - deto + dp * n;

  const auto p_theta = p + theta * dp;
  
  dta = (omega * dt - ta * dp) / (omega + theta * dp);
  const auto ta_theta = ta + theta * dta;
  
  const auto Cs = 1 - exp(-P2 * p_theta * pow(ta_theta, nta));
  const auto Ra = P1 * Cs;
  const auto R = R0 + Q * (1 - exp(-b * p_theta)) + H * p_theta + Ra;
  const auto f = max(seq - R, stress{0});
  fp = fp - de * sinh(f / sinh0) * dt;
}

@UpdateAuxiliaryStateVariables {
  ta = ta + dta;
}
```

My first question is about the way you added a new auxiliary state variable for the aging time "ta". I don't understand the "\theta*dp" in the denominator of the expression of "dta". Because if I write the corresponding incremental equation to (5.7) I get:
![image](https://github.com/thelfer/tfel/assets/65529689/aaf26c4e-5293-4ca6-b3c6-54e7e7c781b0)

Moreover, in all the other equations as for "dp" the state variable is udpated before computing "fp" while here "ta" is updated after computing "dta". Actually I am not sure to perfeclty understand how does an auxialiary state variable work. Indeed, so the update with respect to "theta" inside the integrator is to be able to use the updated value at the current time step while the "@UpdateAuxiliaryStateVariables" outside the integrator is to update "ta" for the next time step after leaving the integrator ? Would it be possible to consider "ta" as another state variable as "p" ?
This is what I tried with the following code inspirated from the documentation and the example of the Chaboche law where a new state varibale "a" is introduced:
```
@DSL Implicit;
@Behaviour EKMC;

@Algorithm NewtonRaphson_NumericalJacobian;
@PerturbationValueForNumericalJacobianComputation 1e-8;
@Epsilon 1e-14;
@Theta 1;

@Brick StandardElasticity;

@MaterialProperty strainrate de0;
@MaterialProperty real Ea;
@MaterialProperty real Va;
@MaterialProperty stress P1;
@MaterialProperty real P2;
@MaterialProperty real omega;
@MaterialProperty stress R0;
@MaterialProperty stress Q;
@MaterialProperty real b;
@MaterialProperty stress H;
@MaterialProperty real nta;
@MaterialProperty real alpha;

@StateVariable strain p;
p.setGlossaryName("EquivalentViscoplasticStrain");

@StateVariable time ta;

@LocalVariable strainrate de;

@LocalVariable stress sinh0;

@InitLocalVariables {
  constexpr auto kB = PhysicalConstants::kb;
  const auto T_theta = T + theta * dT;
  de = de0 * exp(-Ea / (kB * (T_theta)));
  sinh0 = kB * T_theta / Va;
}

@Integrator {
  const auto seq = sigmaeq(sig);
  Stensor n = Stensor(0);
  if(seq > 1.e-12){
    n = 1.5 * deviator(sig)/seq;
  }

  feel = deel - deto + dp * n;

  const auto p_theta = p + theta * dp;
  
  const auto ta_theta = ta + theta * dta;
  fta = fta - dt + dp/omega * ta;
  
  const auto Cs = 1 - exp(-P2 * p_theta * pow(ta_theta, nta));
  const auto Ra = P1 * Cs;
  const auto R = R0 + Q * (1 - exp(-b * p_theta)) + H * p_theta + Ra;
  const auto f = max(seq - R, stress{0});
  fp = fp - de * sinh(f / sinh0) * dt;
}
```
Nevertheless, It is not converging:
![image](https://github.com/thelfer/tfel/assets/65529689/118e8ace-b124-404a-b5f5-3cbf341e041c)

So, there is probably something I did not get with the state variable and the way to integrate it.

Thank you for your help

By the way, here is the mtest code:
```
@Behaviour<umat> './src/libUmatBehaviour.dll' 'umatekmc' ;

@AccelerationAlgorithm 'Cast3M';

@MaterialProperty<constant> 'YoungModulus' 210e9;
@MaterialProperty<constant> 'PoissonRatio' 0.3;
@MaterialProperty<constant> 'MassDensity' 0.0 ;
@MaterialProperty<constant> 'ThermalExpansion' 0.0 ;
@MaterialProperty<constant> 'ReferenceTemperatureForThermalExpansionCoefficient' 0.0 ;
@MaterialProperty<constant> 'ReferenceTemperatureForThermalExpansion' 0.0 ;

@MaterialProperty<constant> 'R0' 200e6;
@MaterialProperty<constant> 'Q' 270e6;
@MaterialProperty<constant> 'b' 23;
@MaterialProperty<constant> 'H' 440e6;
@MaterialProperty<constant> 'de0' 1e-7;
@MaterialProperty<constant> 'Ea' 3.52e-21;
@MaterialProperty<constant> 'Va' 1527e-30;
@MaterialProperty<constant> 'nta' 0.66;
@MaterialProperty<constant> 'P1' 120e6;
@MaterialProperty<constant> 'omega' 0.0015;
@MaterialProperty<constant> 'P2' 3e-4;
@MaterialProperty<constant> 'alpha' 0.1;

@ExternalStateVariable 'Temperature' 293.15 ;

@ImposedStrain 'EXX' { 0. :  0., 30. : 0.02};
@Times {0.,30. in 200
```

**Comment by thelfer** [comment]
> I don't understand the "\theta*dp" in the denominator of the expression of "dta"

Equation 1 leads to:

dta/dt = 1-dp/dt/w*(ta+theta*dta) 

Then you solve for dta ..

**Reply by RonanRiverie** [reply]
Ok, yes indeed. And I suppose that the way I did is because first I update ta to ta+theta*dta and then I use this updated value to compute the new value of dta. But doing so can only workwhen we want to solve for dta which is not the case with your method and the use of an auxiliary variable since you directly and explicitly find dta corresponding to your new time step.

**Comment by thelfer** [comment]
For auxiliary state variables, see https://thelfer.github.io/tfel/web/implicit-dsl.html.

We declare ta as an auxiliary state variables to eliminate dta from the list of unknowns. But if we do that, we need to explicitly update ta after the convergence of the Newton algorithm

**Reply by RonanRiverie** [reply]
Ok I see thanks.

**Comment by thelfer** [comment]
This line is wrong:  

```cxx
  const auto ta_theta = ta + theta * dta;
  fta = fta - dt + dp/omega * ta;
```

You must write:

```cxx
  const auto ta_theta = ta + theta * dta;
  fta = dta - dt + dp/omega * ta_theta;
```

**Reply by RonanRiverie** [reply]
I yes indeed. So by the way it was also "dp" instead of "fp" in this line " fp = fp - de * sinh(f / sinh0) * dt". Actually "fp" is just the name of the equation to solve for the increment "dp" with the Newton Algorithm.

So I changed that line but I am still getting the same error:
![image](https://github.com/thelfer/tfel/assets/65529689/594b4b47-72c7-4663-a2c1-c91538d5c85b)

**Reply by thelfer** [reply]
>  So by the way it was also "dp" instead of "fp" 

It works because fp is initialized to dp. But it is more a mistake than anything else..

**Reply by RonanRiverie** [reply]
Yes exactly and actually for the equation with the variable "ta" there is no need to use a Newton algorithm to solve it since we can just solve it explicitly directly but I was just wondering how to do by declaring it as a state variable to better understand mfront.

**Reply by thelfer** [reply, marked as answer]
This seems to help:

const auto ta_theta = max(ta + theta * dta, time(0));

You must check that ta and fta have the same order of magnitude than dp..

**Reply by RonanRiverie** [reply]
Thanks, It works perfectly:
![image](https://github.com/thelfer/tfel/assets/65529689/35cfb523-919f-4768-ad72-9d2fd6965b05)

So with this line we prevent the aging time to become negative?
Sorry, I don't get the link with dp?

**Reply by thelfer** [reply]
> So with this line we prevent the aging time to become negative?

Exactly. That avoids "NaN" (Not a Number) when calling the `pow` function

> Sorry, I don't get the link with dp?

For efficiency, the Implicit DSL assumes that all unknowns (eel, fp, ta) and residuals (feel, fp, fta) have the same order of magnitudes. Otherwise, we can run in numerical troubles

**Reply by RonanRiverie** [reply]
Ok, I see. Thank you very much once again for your answers. Now I will try to use it into CAST3M to model first a simple tensile test.
